### PR TITLE
Do not create unused testCluster

### PR DESCRIPTION
**Pull request by breskeby**
This avoids creating test clusters that are not required during the build.
We use lazy configuration here on testClusters and only instantiate them as required.

Also updated BUILDING.MD giving hints on using test cluster plugin properly

### CODE REVIEW

File: qa/ccs-rolling-upgrade-remote-cluster/build.gradle [hunk 1]
```diff
@@ -20,8 +20,8 @@ dependencies {
 }
 
 for (Version bwcVersion : BuildParams.bwcVersions.wireCompatible) {
-  String baseName = "v${bwcVersion}"
```
Comment (breskeby) [inline qa/ccs-rolling-upgrade-remote-cluster/build.gradle:23]: This solves a problem with for loops and lazy evaluated strings that causes late resolved GStrings to only contain the last bwcVersion of the for loop

**Review by pugnascotia — APPROVED** [review]

**Comment by breskeby** [comment]
@elasticmachine run elasticsearch-ci/part-2

File: qa/ccs-rolling-upgrade-remote-cluster/build.gradle (hunk 1 above)
Comment (mark-vieira) [reply qa/ccs-rolling-upgrade-remote-cluster/build.gradle:23]: Ug, I've been bitten by that so many times.

File: x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle [hunk 2]
```diff
@@ -53,11 +50,11 @@ tasks.register("writeJavaPolicy") {
 
 tasks.register("follow-cluster", RestIntegTestTask) {
   dependsOn 'writeJavaPolicy', "leader-cluster"
-    useCluster testClusters."leader-cluster"
+    useCluster leaderCluster.get()
```
Comment (mark-vieira) [inline x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:53]: Can we provide an overloaded version of `useCluster` on `TestClusterAware` that accepts a `Provider` so we can avoid all these calls to `get()`?

**Review by mark-vieira — APPROVED** [review]
LGTM, although I wish using lazy configured domain objects didn't introduce so much awkwardness in our build scripts.

File: x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle (hunk 2 above)
Comment (breskeby) [reply x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:53]: could point. I'll give the testclusters api bit more attention

Comment (breskeby) [reply x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle]: I did some polishing on the api. Besides allow passing a task provider here I also added support for Provider in nonSystemProperties. this makes configurations based on clusters more straight forward by allowing using provider mapping: e.g.

```
tasks.register("someRestIntegTest", StandaloneRestIntegTestTask) {
    useCluster baseCluster
    nonInputProperties.systemProperty('tests.rest.cluster', baseCluster.map(c->c.allHttpSocketURI.join(",")))
}
```

Comment (mark-vieira) [reply x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle]: Nice!

**Pull request merged**